pkg/messaging/mqtt/pubsub.go

```diff
@@ -207,7 +223,7 @@ func (sub subscription) indexOf(element string) int {
 }
 
 // Deletes a topic from the slice
-func (sub subscription) delete(topic string) bool {
+func (sub *subscription) delete(topic string) bool {
```

Comment:
While you fixed the problem of modifying the subscription from this method by attaching it to the pointer rather than the struct, have in mind that this does not exactly solve the problem. Here:
```go
func (ps *pubsub) Unsubscribe(id, topic string) error {
	if id == "" {
		return ErrEmptyID
	}
	if topic == "" {
		return ErrEmptyTopic
	}
	ps.mu.Lock()
	defer ps.mu.Unlock()
	// Check client ID
	s, ok := ps.subscriptions[id]
	switch ok {
	case true:
		// Check topic
		if ok := s.contains(topic); !ok {
			return ErrNotSubscribed
		}
	default:
		return ErrNotSubscribed
	}
	token := s.client.Unsubscribe(topic)
	if token.Error() != nil {
		return token.Error()
	}

	ok = token.WaitTimeout(ps.timeout)
	if !ok {
		return ErrUnsubscribeTimeout
	}
	if ok := s.delete(topic); !ok {
		return ErrUnsubscribeDeleteTopic
	}
	if len(s.topics) == 0 {
		delete(ps.subscriptions, id)
	}
	return token.Error()
}
```
You call it on `s` where `s` is `s, ok := ps.subscriptions[id]` which means that you are effectively calling it on the **copy** of the value in the map and all the changes on `s` will not be reflected in `ps.subscriptions[id]`.

Reply:
```
if ok := s.delete(topic); !ok {
		return errUnsubscribeDeleteTopic
		return ErrUnsubscribeDeleteTopic
	}
	ps.subscriptions[id] = s

	if len(s.topics) == 0 {
		delete(ps.subscriptions, id)
	}
	return token.Error()
```

Is this okay, since it updates the map attached to pubsub, and later len(s.topics) is equivalent to len(ps.subscriptions)

Reply:
It is, but probably the simplest solution would be to use pointers in `subscriptions`  map like:
```go
type pubsub struct {
	publisher
	logger        log.Logger
	mu            *sync.RWMutex
	address       string
	timeout       time.Duration
	subscriptions map[string]*subscription
}
```

Reply:
@dborovcanin , I tested with converting map to pointer, and the complexity remains mostly the same. With more added confusion of maintaining pointers. (And sometimes I get nil pointer dereference panic also)

So, I feel like keeping the current format is the better approach (Since it only adds 2 lines to the program) (where we reassign `subscriptions[id] =s` , to update the original map)